packages/devtools_app/lib/src/shared/diagnostics/inspector_service.dart

```diff
@@ -418,12 +418,53 @@ class InspectorService extends InspectorServiceBase {
     return false;
   }
 
+  /// Ensures that addPubRootDirectories, getPubRootDirectories, and
+  /// removePubRootDirectories have all registered as services
+  /// before completing.
+  Future<void> waitUntilPubRootDirectoriesServicesAvailable() {
+    final futureListener = Completer();
+    bool addAvailable = false;
+    bool getAvailable = false;
+    bool removeAvailable = false;
+    void checkIfComplete() {
+      if (addAvailable && getAvailable && removeAvailable) {
+        futureListener.complete();
+      }
+    }
+
+    serviceConnection.serviceManager
+        .registeredServiceListenable(
+            WidgetInspectorServiceExtensions.addPubRootDirectories.name)
+        .addListener(() {
+      addAvailable = true;
+      checkIfComplete();
+    });
+    serviceConnection.serviceManager
+        .registeredServiceListenable(
+            WidgetInspectorServiceExtensions.getPubRootDirectories.name)
+        .addListener(() {
+      getAvailable = true;
+      checkIfComplete();
+    });
+    serviceConnection.serviceManager
+        .registeredServiceListenable(
+            WidgetInspectorServiceExtensions.removePubRootDirectories.name)
+        .addListener(() {
+      removeAvailable = true;
+      checkIfComplete();
+    });
+
+    return futureListener.future.timeout(const Duration(seconds: 15));
+  }
+
   Future<void> addPubRootDirectories(List<String> rootDirectories) async {
+    await waitUntilPubRootDirectoriesServicesAvailable();
```

Comment:
what if we hit the timeout, we would not want to call the service if it is not available

Reply:
It looks like timeout throws a TimeoutException so I don't think it would continue running